drivers: can: common: fix sample point calculations

henrikbrixandersen opened the pull request:
- Always include the SYNC segment of 1 time quanta when calculating ts1 and ts2.
- Respect the range limits of phase_seg1 when attempting to distribute ts1 between prop_seg and phase_seg1. Even distribution may not be possible if the allowed ranges of prop_seg and phase_seg1 are not equal.

Fixes: #55919

henrikbrixandersen committed:
drivers: can: common: include sync segment in timing calculation

Always include the SYNC segment of 1 time quanta when calculating ts1 and
ts2.

Fixes: #55919

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: common: respect the range limits of phase_seg1

Respect the range limits of phase_seg1 when attempting to distribute ts1
between prop_seg and phase_seg1.

Even distribution may not be possible if the allowed ranges of prop_seg and
phase_seg1 are not equal.

Fixes: #55919

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

martinjaeger commented:
I'm trying to wrap my head around it, but I'm having a hard time to understand what `ts`, `ts1` and `ts2` mean exactly. Some more documentation/comments for these internal functions would really be helpful, IMO.

Do you think you could add some explanatory comments or rename the variables to something more obvious to help with proper review? As far as I can see, `TS`, `TS1` and `TS2` are not typical unambiguous names used for CAN bit timing calculations.

henrikbrixandersen committed:
drivers: can: common: document sample point calculation function

Add documentation for the update_sampling_pnt() function and rename a few
internal variables to improve readability.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
> I'm trying to wrap my head around it, but I'm having a hard time to understand what `ts`, `ts1` and `ts2` mean exactly. Some more documentation/comments for these internal functions would really be helpful, IMO.
> 
> Do you think you could add some explanatory comments or rename the variables to something more obvious to help with proper review? As far as I can see, `TS`, `TS1` and `TS2` are not typical unambiguous names used for CAN bit timing calculations.

A fair request - I was also a bit lost trying to debug it. I want to follow up general clean-up/documentation PR for `can_common.c` but I have added documentation and renaming of variables for `update_sampling_pnt()` here.

martinjaeger reviewed (dismissed):
Thanks for the additional documentation. Now I can understand it and could not spot any mistake.

str4t0m commented on drivers/can/can_common.c:
```diff
@@ -105,9 +105,17 @@ static int update_sampling_pnt(uint32_t ts, uint32_t sp, struct can_timing *res,
 	res->phase_seg1 = ts1 - res->prop_seg;
-	res->phase_seg2 = ts2;
+
+	if (res->phase_seg1 > max->phase_seg1) {
```
Theoretically if `phase_seg1` would have a min value greater than the min value of `prop_seg` it could still lead to an invalid result for `phase_seg1`.

henrikbrixandersen replied:
@str4t0m You are right, thank you for spotting this. I had somehow convinced myself this check was already covered by the clamping of the propagation segment, but clearly it isn't. I have update the code to also check the minimum boundary of `phase_seg1`.

str4t0m reviewed (commented):
Thanks for these Fixes and improved documentation.
